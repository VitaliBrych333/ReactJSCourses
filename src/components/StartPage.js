import React, { Component, Fragment } from 'react';
import SearchHeader from './SearchHeader';
import NotFound from './NotFound';
import styled from 'styled-components'
import { connect } from 'react-redux';
import CardFilm from './CardFilm';

const StyledSection = styled.section`
    padding: 25px;
    background-color:  rgb(209, 189, 189);
    display: flex;
    flex-wrap: wrap;
    justify-content: space-around;

    div {
        margin-bottom: 10px;
    }
`;

class StartPage extends Component {
    constructor(props) {
        super(props)
    }

    render() {
        const { error, loading, data } = this.props;
        let main;

        if (this.props.data.length) {
            main =  <StyledSection>
                        {data.map(item => <CardFilm key={item.id} info={item}/> )}
                    </StyledSection>
        } else {
            main = <NotFound/>;
        }

        return (
            <Fragment>
                <SearchHeader count={this.props.total}/>
                {main}
            </Fragment>
        );
    }
};

const mapStateToProps = state => ({
    data: state.movieReducer.movies.data,
    total: state.movieReducer.movies.total,
    loading: state.movieReducer.loading,
    error: state.movieReducer.error,
});

export default connect(mapStateToProps)(StartPage);
