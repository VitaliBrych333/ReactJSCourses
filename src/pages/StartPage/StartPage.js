import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import queryString from 'query-string';
import CardFilm from '../../components/CardFilm';
import EditPage from '../../components/EditPage';
import DeleteWindow from '../../components/DeleteWindow';
import SearchHeader from '../../components/SearchHeader';
import NotFound from '../../components/NotFound';
import {
  fetchMovies,
  fetchMoviesByGenre,
  fetchMoviesBegin,
} from '../../redux/actions/moviesActions';

const StyledSection = styled.section`
  padding: 25px;
  background-color: rgb(209, 189, 189);
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;

  .card {
    margin-top: 15px;
    margin-bottom: 10px;
  }
`;

class StartPage extends PureComponent {
  static initialAction(req) {
    const { sortBy, search, searchBy, filter } = req.query;
    if (sortBy && search) {
      return fetchMovies(sortBy, search);
    }

    if (sortBy && searchBy && filter) {
      return fetchMoviesByGenre(sortBy, filter);
    }

    return fetchMoviesBegin();
  }

  componentDidUpdate(prevProps) {
    const { fetchMovies, fetchMoviesByGenre, location } = this.props;

    if (prevProps.location.search !== location.search) {
      const url = location.search;
      const params = queryString.parse(url);
      const { sortBy, search, searchBy, filter } = params;

      if (sortBy && search) {
        fetchMovies(sortBy, search);
      }

      if (sortBy && searchBy && filter) {
        fetchMoviesByGenre(sortBy, filter);
      }
    }
  }

  render() {
    const { data, isShowEditPage, isShowDeletePage } = this.props;
    let main;

    if (data !== null && data.length) {
      main = (
        <StyledSection>
          {data.map((item) => (
            <CardFilm key={item.id} info={item} />
          ))}
        </StyledSection>
      );
    } else if (data !== null && !data.length) {
      main = <NotFound />;
    }

    return (
      <>
        {isShowEditPage && <EditPage />}
        {isShowDeletePage && <DeleteWindow />}
        <SearchHeader />
        {main}
      </>
    );
  }
}

StartPage.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  isShowEditPage: PropTypes.bool,
  isShowDeletePage: PropTypes.bool,
  fetchMovies: PropTypes.func,
  fetchMoviesByGenre: PropTypes.func,
  location: PropTypes.objectOf(PropTypes.any),
};

const mapDispatchToProps = {
  fetchMovies,
  fetchMoviesByGenre,
};

const mapStateToProps = (state) => ({
  data: state.movieReducer.moviesByCriteria.data,
  isShowEditPage: state.windowReducer.isShowEditPage,
  isShowDeletePage: state.windowReducer.isShowDeletePage,
});

export default connect(mapStateToProps, mapDispatchToProps)(StartPage);
