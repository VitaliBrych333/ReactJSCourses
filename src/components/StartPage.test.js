import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow } from 'enzyme';
import StartPage from './StartPage';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

Enzyme.configure({ adapter: new Adapter() });

describe('<StartPage/>', () => {
    const initialState = {
        movieReducer: {
            movies: {
                data: [],
                total: 0
            }
        },
        criteriaReducer: {
            search: 'search',
            sort: 'sort'
        },
        windowReducer: {
            showModal: false,
            showEditPage: false,
            showDeletePage: false,
            showAddPage: false
        }
    };
    const mockStore = configureStore();
    let store, wrapper;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        store = mockStore(initialState);
        wrapper = shallow(<Provider store={store}><StartPage /></Provider>);
    });

    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it('render component', () => {
        expect(wrapper.find(StartPage).length).toEqual(1);
    });

    it('should equals to snapshot of CriteriaSearch', () => {
        const renderedValue = renderer.create(<Router><Provider store={store}><StartPage /></Provider></Router>).toJSON();
        expect(renderedValue).toMatchSnapshot();
    });
})
