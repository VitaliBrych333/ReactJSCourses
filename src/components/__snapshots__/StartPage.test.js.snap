// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StartPage/> should equals to snapshot of CriteriaSearch 1`] = `
Array [
  <header
    className="sc-bwzfXH kfyadn"
  >
    <h1>
      Find your movie
    </h1>
    <div
      className="sc-bdVaJa lmvhs mb-3 input-group"
    >
      <input
        aria-describedby="basic-addon2"
        aria-label="Recipient's username"
        className="form-control"
        onChange={[Function]}
        placeholder="Please write the film name"
      />
      <div
        className="input-group-append"
      >
        <button
          className="btn btn-outline-danger"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
      </div>
    </div>
    <p>
      Search
       by
    </p>
    <div
      aria-label="Basic example"
      className="btn-group"
      role="group"
    >
      <button
        className="btn btn-danger"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
      <button
        className="btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Genre
      </button>
    </div>
    <div
      className="count-movie"
    >
      <p>
        Sort
         by
      </p>
      <div
        aria-label="Basic example"
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-danger"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Release date
        </button>
        <button
          className="btn btn-secondary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Rating
        </button>
      </div>
    </div>
  </header>,
  <h3
    className="sc-htpNat hNvbdx"
  >
    No films found
  </h3>,
]
`;
